src/slice/sort.cr

```diff
@@ -206,4 +206,357 @@ struct Slice(T)
+      while r = collapse(runs)
+        left = runs[r + 1]
+        right = runs[r]
+        merge!(v[left.begin...right.end], left.size, buf)
+        runs[r] = left.begin...right.end
+        runs.delete_at(r + 1)
+      end
+    end
+  end
+
+  # Inserts `v[0]` into pre-sorted sequence `v[1..]` so that whole `v[..]` becomes sorted.
+  #
+  # This is the integral subroutine of insertion sort.
+  protected def self.insert_head!(v)
+    if v.size >= 2 && cmp(v[1], v[0]) < 0
+      x, v[0] = v[0], v[1]
+      (2...v.size).each do |i|
+        if cmp(v[i], x) < 0
+          v[i - 1] = v[i]
+        else
+          v[i - 1] = x
+          return
+        end
+      end
+      v[v.size - 1] = x
+    end
+  end
+
+  # Merges non-decreasing runs `v[..mid]` and `v[mid..]` using `buf` as temporary storage, and
+  # stores the result into `v[..]`.
+  protected def self.merge!(v, mid, buf)
+    size = v.size
+
+    if mid <= size - mid
+      # The left run is shorter.
+      buf.copy_from(v.to_unsafe, mid)
+
+      left = 0
+      right = mid
+      out = v.to_unsafe
+
+      while left < mid && right < size
+        # Consume the lesser side.
+        # If equal, prefer the left run to maintain stability.
+        if cmp(v[right], buf[left]) < 0
+          out.value = v[right]
+          out += 1
+          right += 1
+        else
+          out.value = buf[left]
+          out += 1
+          left += 1
+        end
+      end
+
+      out.copy_from(buf + left, mid - left)
+    else
+      # The right run is shorter.
+      buf.copy_from((v + mid).to_unsafe, size - mid)
+
+      left = mid
+      right = size - mid
+      out = v.to_unsafe + size
+
+      while left > 0 && right > 0
+        # Consume the greater side.
+        # If equal, prefer the right run to maintain stability.
+        if cmp(buf[right - 1], v[left - 1]) < 0
+          left -= 1
+          out -= 1
+          out.value = v[left]
+        else
+          right -= 1
+          out -= 1
+          out.value = buf[right]
+        end
+      end
+
+      (v + left).copy_from(buf, right)
+    end
+  end
+
+  # This merge sort borrows some (but not all) ideas from TimSort, which is described in detail
+  # [here](http://svn.python.org/projects/python/trunk/Objects/listsort.txt).
+  #
+  # The algorithm identifies strictly descending and non-descending subsequences, which are called
+  # natural runs. There is a stack of pending runs yet to be merged. Each newly found run is pushed
+  # onto the stack, and then some pairs of adjacent runs are merged until these two invariants are
+  # satisfied:
+  #
+  # 1. for every `i` in `1..runs.len()`: `runs[i - 1].len > runs[i].len`
+  # 2. for every `i` in `2..runs.len()`: `runs[i - 2].len > runs[i - 1].len + runs[i].len`
+  #
+  # The invariants ensure that the total running time is `O(n * log(n))` worst-case.
+  protected def self.merge_sort!(v : Slice(T), comp) forall T
+    size = v.size
```

Comment:
`v` is not very descriptive, `other` will be better.

`size` is already a method of `Slice`, a better name would be `other_size`.

Reply:
The name `v` comes from Rust. I guess it means "**v**ector" which is slice of Crystal in Rust.

And, this variable refers to the slice to be sorted, so I think `other` does not make sense. We can rename it to `slice` for Crystal.

Reply:
Please let's use the names as in Rust, it'll be easier to apply patches.

Also, feel free to ignore comments unless they are from a core team member.

Reply:
> Also, feel free to ignore comments unless they are from a core team member.

That sounds a bit harsh, doesn't it? Reviews and suggestions from everyone are welcome, but it's usually advisable to wait for other comments, especially from core team members (because that's the ones who have to approve a PR).

Reply:
Truth sometimes is harsh, yes.

Reply:
Well, the truth is that some comments don't need to be addressed yet people don't know that and they lose time reverting or discussing those things. And some are simply plain wrong and I have to lose time jumping to address them to avoid others wasting time. But I guess it's better if I don't lose my time plus I don't offend others.

Reply:
@asterite said in this thread before:

> Please let's use the names as in Rust, it'll be easier to apply patches.

This was removed, but I agree **this** opinion really. Anyone, what do you think?

Reply:
I didn't have this insight. I even tought it can be a private instance method, but that's right, better let it a pure function similar to the Rust one 👍